src/details/ArborX_DetailsFDBSCAN.hpp

```diff
@@ -46,53 +46,52 @@ struct CountUpToN
 template <typename MemorySpace, typename CorePointsType>
 struct FDBSCANCallback
 {
-  UnionFind<MemorySpace> union_find_;
-  CorePointsType is_core_point_;
+  UnionFind<MemorySpace> _union_find;
+  CorePointsType _is_core_point;
 
   FDBSCANCallback(Kokkos::View<int *, MemorySpace> const &view,
                   CorePointsType is_core_point)
-      : union_find_(view)
-      , is_core_point_(is_core_point)
+      : _union_find(view)
+      , _is_core_point(is_core_point)
   {
   }
 
   template <typename Query>
-  KOKKOS_FUNCTION void operator()(Query const &query, int j) const
+  KOKKOS_FUNCTION auto operator()(Query const &query, int j) const
   {
     int const i = ArborX::getData(query);
 
-    // NOTE: for halo finder/ccs algorithm (in which is_core_point(i) is always
-    // true), the algorithm below will be simplified to
-    //   if (i > j)
-
-    if (!is_core_point_(j))
+    bool const is_border_point = !_is_core_point(i);
+    if (is_border_point)
     {
-      // The neighbor is not a core point, do nothing
-      return;
+      // Ignore border points, they will be processed by the
+      // connected core points. Theoretically, border points could have been
+      // filtered out prior to running the algorithm, but that may be expensive.
+      return ArborX::CallbackTreeTraversalControl::early_exit;
     }
 
-    bool is_border_point =
-        !is_core_point_(i); // is_core_point_(j) is aready true
-
-    if (is_border_point && union_find_.representative(i) == i)
+    bool const is_neighbor_core_point = _is_core_point(j);
```

Comment:
I am not a big fan of that variable name.  I think is reads poorly.  What about `neighbor_is_core_point`?

Reply:
Prefer booleans to start with `is_blah`. Not fixing. One thing that may need fixing, though, is `is_border_point`, because that one is misleading (it could be border **or** noise point).

Reply:
> Prefer booleans to start with is_blah. Not fixing.

For the record, I think the name is misleading.